llvm/unittests/Object/ELFObjectFileTest.cpp

```diff
@@ -897,6 +897,34 @@ TEST(ELFObjectFileTest, InvalidDecodePGOAnalysisMap) {
 
   DoCheck(MissingBrProb, "unable to decode LEB128 at offset 0x00000017: "
                          "malformed uleb128, extends past end");
+
+  // Check that we fail when pgo data exists but the feature bits are disabled.
+  SmallString<128> ZeroFeatureButWithAnalyses(CommonYamlString);
+  ZeroFeatureButWithAnalyses += R"(
+        Version: 2
+        Feature: 0x00
+        BBEntries:
+          - ID:            1
+            AddressOffset: 0x0
+            Size:          0x1
+            Metadata:      0x6
+          - ID:            2
+            AddressOffset: 0x1
+            Size:          0x1
+            Metadata:      0x2
+      - Address: 0x11111
+        Version: 2
+        Feature: 0x00
+        BBEntries: []
+    PGOAnalyses:
+      - PGOBBEntries:
+         - BBFreq:        1000
+         - BBFreq:        1000
+      - PGOBBEntries: []
+)";
+
+  DoCheck(ZeroFeatureButWithAnalyses,
+          "unsupported SHT_LLVM_BB_ADDR_MAP version: 232");
```

Comment:
That's only the case if `len(PGOAnalyses) != len(Entries)` which isn't the case here.

I could remove those returns in ELFEmitter so that it just emits Entries and PGOAnalyses interspersed until the shorter one runs out then just emit the longer one until the end. Do you think this is preferable for testing?

Reply:
I noticed that this is the case on the previous test MissingBrProb. I'm going to remove the early returns so that ELFEmitter is more geared for testing.

Reply:
I updated ELFYaml emitting so that we can emit invalid forms for testing. The emitter will now just emit Entries and PGOAnalyses interspersed until one runs out in which is just emits the other one until it ends.

Kept in the warnings in ELFYaml. I also added a two more tests.

Reply:
My bad. I misinterpreted the if conditions in ELFEmitter.cpp (I thought they were checking the feature value).

On a different note, I think we should make these lit tests and not unit tests. As we add more unit tests in a single file, maintenance/reviews get harder. WDYT?

Reply:
If these were to become lit tests, what tool would be best since these tests use the direct Object api? Maybe elf2yaml and readobj, but I haven't implemented the support for PGO Analysis Map in them yet.

Reply:
You only need yaml2obj to create these tests. Two things must be tested. 
1. How yaml2obj creates the section. 
2. How llvm-readobj reads and decodes the created section.

You can hold on to the patch if you want to do this at a later time.

Reply:
Sounds good, I'll put this to drafts and either revisit or make it a part of a separate PR.

Reply:
FWIW, I would lean towards unit tests if the setup is straightforward. Apart from anything else, lit tests on Windows have a non-trivial performance cost (much worse than extending an existing unit test), due to the process launching overhead.

Regardless of the approach taken, it's important to test all failure cases in the ObjectYAML code, plus cases like "this thing is optional - test that it can be omitted or specified" and "this thing is required - test that an error is emitted if missing".

Reply:
> FWIW, I would lean towards unit tests if the setup is straightforward. Apart from anything else, lit tests on Windows have a non-trivial performance cost (much worse than extending an existing unit test), due to the process launching overhead.

Thanks for chiming in @jh7370 . One problem I have with our unit tests is that if we test multiple cases in a single test function, it's very hard to associate a failure to the test case. Although, while writing this I realized we can extend our test lambda (`DoCheckFails` and `DoCheckSucceeds`) with a string to represent the test case.

Reply:
On the note of tracing back errors to test cases, some of the clang-format unit tests have a pattern where they take a source location as arguments, then provide a macro to automatically put in `__FILE__` and `__LINE__` macros to make it easier to trace back errors. I'll try to implement that in a separate PR and then possibly reopen this if we find that approach good.